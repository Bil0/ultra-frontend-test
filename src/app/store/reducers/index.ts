import { MetaReducer } from '@ngrx/store';
import { environment } from 'environments/environment';

import * as fromSession from './marketplace.reducer';

export interface AppState {
  readonly marketplace: fromSession.MarketplaceState;
}

export const metaReducers: MetaReducer<AppState>[] = !environment.production
  ? []
  : [];
