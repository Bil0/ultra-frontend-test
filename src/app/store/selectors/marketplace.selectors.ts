import { createFeatureSelector, createSelector } from '@ngrx/store';
import { HomeProduct, Product } from 'app/models/product.model';

import * as fromMarketplace from '../reducers/marketplace.reducer';

export const selectMarketplaceState =
  createFeatureSelector<fromMarketplace.MarketplaceState>('marketplace');

export const selectWallet = createSelector(
  selectMarketplaceState,
  (state: fromMarketplace.MarketplaceState) => state.wallet
);

export const selectProducts = createSelector(
  selectMarketplaceState,
  (state: fromMarketplace.MarketplaceState) => state.products
);

export const selectProductIds = createSelector(
  selectMarketplaceState,
  (state: fromMarketplace.MarketplaceState) => state.productIds
);

export const selectProductIdsFromBasket = createSelector(
  selectMarketplaceState,
  (state: fromMarketplace.MarketplaceState) => state.basket
);

export const selectHomeProducts = createSelector(
  selectProductIds,
  selectProducts,
  selectProductIdsFromBasket,
  (ids, products, productIdsFromBasket): HomeProduct[] => {
    return ids.map((id) => ({
      ...(products[id] as Product), // we'll just assume that every field you access is valid
      isInBasket: productIdsFromBasket.includes(id),
    }));
  }
);

export const selectBasket = createSelector(
  selectProducts,
  selectProductIdsFromBasket,
  (products, productIdsFromBasket) => {
    const basketProducts = productIdsFromBasket.map(
      (id) => products[id] as Product
    );

    return {
      products: basketProducts,
      total: basketProducts.reduce((prev, product) => prev + product.price, 0),
    };
  }
);

export const selectBasketTotalItems = createSelector(
  selectProductIdsFromBasket,
  (ids: number[]) => ids.length
);
