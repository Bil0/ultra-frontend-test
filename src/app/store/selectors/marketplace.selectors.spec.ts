import { createMarketplaceState } from 'app/testing/store.mock';
import * as fromMarketplace from './marketplace.selectors';

describe('Marketplace Selectors', () => {
  const testCases = [
    {
      name: 'select MarketplaceState',
      selector: fromMarketplace.selectMarketplaceState,
      state: createMarketplaceState(),
    },
    {
      name: 'select wallet',
      selector: fromMarketplace.selectWallet,
      state: createMarketplaceState(),
    },
    {
      name: 'select WalletBalance',
      selector: fromMarketplace.selectWalletBalance,
      state: createMarketplaceState(),
    },
    {
      name: 'select ProductIdsFromBasket',
      selector: fromMarketplace.selectProductIdsFromBasket,
      state: createMarketplaceState(),
    },
    {
      name: 'select HomeProducts',
      selector: fromMarketplace.selectHomeProducts,
      state: createMarketplaceState(),
    },
    {
      name: 'select Entities',
      selector: fromMarketplace.selectEntities,
      state: createMarketplaceState(),
    },
  ];

  testCases.forEach(({ name, state, selector }) => {
    it(`Should ${name}`, () => {
      expect(selector.projector(state)).toMatchSnapshot();
    });
  });
});
