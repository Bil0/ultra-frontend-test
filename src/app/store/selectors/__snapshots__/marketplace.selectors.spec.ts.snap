// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marketplace Selectors Should select Entities 1`] = `Object {}`;

exports[`Marketplace Selectors Should select HomeProducts 1`] = `Array []`;

exports[`Marketplace Selectors Should select MarketplaceState 1`] = `
Object {
  "basket": Array [],
  "entities": Object {},
  "ids": Array [],
  "wallet": Object {
    "balance": 100,
    "id": 1,
    "userId": 1,
  },
}
`;

exports[`Marketplace Selectors Should select ProductIdsFromBasket 1`] = `Array []`;

exports[`Marketplace Selectors Should select WalletBalance 1`] = `undefined`;

exports[`Marketplace Selectors Should select wallet 1`] = `
Object {
  "balance": 100,
  "id": 1,
  "userId": 1,
}
`;
